Python/Product/PythonTools/PythonTools/Project/PythonProjectInfoBar.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using Microsoft.PythonTools.Logging;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudioTools;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.PythonTools.Project {
+    internal abstract class PythonProjectInfoBar : IVsInfoBarUIEvents, IDisposable {
+        private readonly IVsShell _shell;
+        private readonly IVsInfoBarUIFactory _infoBarFactory;
+        private IdleManager _idleManager;
+        private uint _adviseCookie;
+        private IVsInfoBarUIElement _infoBar;
+        private InfoBarModel _infoBarModel;
+
+        protected PythonProjectInfoBar(PythonProjectNode projectNode) {
+            Project = projectNode ?? throw new ArgumentNullException(nameof(projectNode));
+            Logger = (IPythonToolsLogger)projectNode.Site.GetService(typeof(IPythonToolsLogger));
+            _shell = (IVsShell)projectNode.Site.GetService(typeof(SVsShell));
+            _infoBarFactory = (IVsInfoBarUIFactory)projectNode.Site.GetService(typeof(SVsInfoBarUIFactory));
+            _idleManager = new IdleManager(Project.Site);
+        }
+
+        protected PythonProjectNode Project { get; }
+
+        protected IPythonToolsLogger Logger { get; }
+
+        protected bool IsCreated => _infoBar != null;
+
+        public abstract Task CheckAsync();
+
+        protected void Create(InfoBarModel model) {
+            _infoBarModel = model;
+            FinishCreate();
+        }
+
+        protected void Close() {
+            _infoBar.Close();
+        }
+
+        protected void FinishCreate() {
+            if (_infoBar != null) {
+                return;
```

Comment:
Is it ok that `_infoBar` references `InfoBarModel` that is different from `_infoBarModel`?

Reply:
It's extra protection that should never get it. If the info bar has been created, it's a programming error to want to create it again. In fact, I'll add an assert.